### PR TITLE
[Fix][Elementwise] Correct maximum/minimum signed-zero tie-break

### LINKED ISSUE #469: [ELEMENTWISE] Restore signed-zero-correct maximum/minimum semantics for fp16/bf16
## Goal
Make `maximum` / `minimum` match PyTorch / IEEE signed-zero behavior while preserving the NaN-propagation fix.

## Context
PR #463 fixed NaN propagation by using:
- `T.if_then_else(a > b, a, b)` for `maximum`
- `T.if_then_else(a < b, a, b)` for `minimum`

That avoids the current TileLang/CUDA half-type overload problem with `T.max` / `T.min`, but it still does **not** preserve signed-zero semantics on equality ties:
- `maximum(+0.0, -0.0)` should return `+0.0`
- `minimum(-0.0, +0.0)` should return `-0.0`

With ordered comparisons, equality falls through to the second operand, so the result depends on operand order instead of IEEE/PyTorch tie behavior.

## Scope
- Reproduce the current signed-zero mismatch for fp16/bf16/fp32.
- Evaluate whether TileLang now supports a half-compatible lowering for ordered `max` / `min`.
- If not, design an alternative lowering that preserves:
  - NaN propagation
  - signed-zero tie behavior
  - acceptable performance
- Add explicit regression tests for signed-zero cases.
- Document the resolution in the relevant PR / benchmark notes.

## Notes
This is intentionally split from PR #463 because it depends on shared kernel semantics and potentially upstream TileLang lowering behavior, not just the 22 subclasses added there.

### PR DESCRIPTION
**Pull request by lcy-seso**
Closes #469

## Summary

Fix `maximum` / `minimum` elementwise ops to match PyTorch / IEEE 754 signed-zero semantics:
- `maximum(+0.0, -0.0)` now correctly returns `+0.0` (was returning `-0.0`)
- `minimum(-0.0, +0.0)` now correctly returns `-0.0` (was returning `+0.0`)

The fix adds a tie-break path using `copysign` comparison that activates only when `a == b`, preserving the existing NaN-propagation behavior from PR #463.

**Dtype support matrix:**

| dtype | supported |
|-------|-----------|
| float16 | yes |
| bfloat16 | yes |
| float32 | yes |

## Structural Compliance

All checks passed.

## Test plan

- [x] `maximum(+0.0, -0.0)` returns `+0.0` for fp16, bf16, fp32
- [x] `minimum(-0.0, +0.0)` returns `-0.0` for fp16, bf16, fp32
- [x] NaN propagation behavior from PR #463 is preserved (not regressed)
- [x] 12 new signed-zero regression tests added and passing
- [x] No performance regression — tie-break path only activates on `a == b`

## Benchmark

**Environment**: H200 GPU, CUDA 12.8, PyTorch 2.9.1+cu128, TileLang 0.1.8

Before/after comparison (`main` vs this branch, fp16):

| Op | Shape | Before Latency (ms) | Before BW (TB/s) | After Latency (ms) | After BW (TB/s) | Delta |
|----|-------|---------------------|-------------------|---------------------|-------------------|-------|
| maximum | (1024, 4096) | 0.02 | 1.18 | 0.02 | 1.18 | 0% |
| maximum | (1024, 10240) | 0.05 | 1.33 | 0.05 | 1.32 | -0.8% |
| maximum | (1024, 20480) | 0.09 | 1.43 | 0.09 | 1.42 | -0.7% |
| minimum | (1024, 4096) | 0.02 | 1.19 | 0.02 | 1.18 | -0.8% |
| minimum | (1024, 10240) | 0.05 | 1.32 | 0.05 | 1.31 | -0.8% |
| minimum | (1024, 20480) | 0.09 | 1.43 | 0.09 | 1.42 | -0.7% |

**Takeaways:**
- No measurable regression — deltas are within noise margin (<1%)
- The `copysign`-based tie-break only activates when `a == b`, which is rare in real workloads
- Existing gap vs PyTorch baseline (42–65% efficiency) is a pre-existing issue tracked in #512

**Benchmark command:**
```bash
PYTHONPATH="$PWD" python -m pytest benchmarks/ops/bench_binary_elementwise.py -v -k "maximum or minimum"
```

## Regression

- 101 tests pass (12 signed-zero + 10 NaN propagation + existing suite)
- Lint clean (`pre-commit run --all-files` passes)
- Signed-zero fix does not alter NaN-propagation codepath — verified by `test_maximum_signed_zero_with_nan` and `test_minimum_signed_zero_with_nan`

---
Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/ops/test_binary_arith.py [hunk 1]
```diff
@@ -657,6 +657,113 @@ def test_minimum_nan_propagation(dtype: torch.dtype) -> None:
...
+    with torch.no_grad():
+        out = op(a, b)
+    # NaN positions must match
+    assert torch.equal(torch.isnan(out), torch.isnan(ref)), (
+        f"NaN positions differ: out={out}, ref={ref}"
+    )
+    # Non-NaN values: sign bits must match too
+    mask = ~torch.isnan(ref)
+    if mask.any():
+        torch.testing.assert_close(out[mask], ref[mask], atol=0, rtol=0)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_binary_arith.py]: Addressed in fa49de6.

Evidence:
- Tests now use mixed NaN/non-NaN inputs: `a = [nan, 1.0, -0.0, 0.0, nan, 3.0]`, `b = [1.0, nan, 0.0, -0.0, -0.0, 2.0]`
- Added `assert mask.any()` guard to ensure non-NaN path always executes
- Explicit sign-bit check via `torch.signbit` comparison on non-NaN values (line 745)

Current HEAD: fa49de6

File: tests/ops/test_binary_arith.py [hunk 2]
```diff
@@ -657,6 +657,113 @@ def test_minimum_nan_propagation(dtype: torch.dtype) -> None:
...
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_binary_arith.py]: Addressed in fa49de6.

Evidence:
- Tests now use mixed NaN/non-NaN inputs: `a = [nan, -0.0, 0.0, 1.0, nan, 2.0]`, `b = [1.0, nan, -0.0, 0.0, 0.0, 3.0]`
- Added `assert mask.any()` guard to ensure non-NaN path always executes
- Explicit sign-bit check via `torch.signbit` comparison on non-NaN values (line 771)

Current HEAD: fa49de6

[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_binary_arith.py]: Addressed in fa49de6.

Evidence:
- Both _with_nan tests now use mixed NaN/non-NaN inputs ensuring the mask.any() block executes
- Explicit sign-bit comparison added via `torch.signbit(out[mask])` vs `torch.signbit(ref[mask])`
- `assert mask.any()` guard prevents silent test pass when all outputs are NaN

Current HEAD: fa49de6

File: tests/ops/test_binary_arith.py [hunk 3]
```diff
@@ -657,6 +657,113 @@ def test_minimum_nan_propagation(dtype: torch.dtype) -> None:
...
+
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_binary_arith.py:749]: Addressed in fa49de6.

Evidence:
- test_maximum_signed_zero_with_nan now uses mixed inputs: `a = [nan, 1.0, -0.0, 0.0, nan, 3.0]`
- `assert mask.any()` guard ensures non-NaN validation always runs
- Sign-bit check: `torch.signbit(out[mask])` vs `torch.signbit(ref[mask])` (line 745-747)

Current HEAD: fa49de6

File: tests/ops/test_binary_arith.py [hunk 4]
```diff
@@ -657,6 +657,113 @@ def test_minimum_nan_propagation(dtype: torch.dtype) -> None:
...
+        f"Signed-zero mismatch: out signs={out_signbits}, ref signs={ref_signbits}"
+    )
+
+
+@SignedZeroFixture
+def test_maximum_signed_zero_with_nan(dtype: torch.dtype) -> None:
+    """Signed-zero fix must not regress NaN propagation."""
+    nan = float("nan")
+    a = torch.tensor([nan, 0.0, -0.0, nan], dtype=dtype, device="cuda")
+    b = torch.tensor([0.0, nan, nan, -0.0], dtype=dtype, device="cuda")
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_binary_arith.py]: Addressed in fa49de6. Tests now use mixed NaN/non-NaN inputs with explicit sign-bit assertions. See commit fa49de6.

File: tests/ops/test_binary_arith.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_binary_arith.py]: Addressed in fa49de6. Tests now use mixed NaN/non-NaN inputs with explicit sign-bit assertions. See commit fa49de6.

File: tests/ops/test_binary_arith.py [hunk 5]
```diff
@@ -657,6 +657,113 @@ def test_minimum_nan_propagation(dtype: torch.dtype) -> None:
...
+    if mask.any():
+        torch.testing.assert_close(out[mask], ref[mask], atol=0, rtol=0)
+
+
+@SignedZeroFixture
+def test_minimum_signed_zero_with_nan(dtype: torch.dtype) -> None:
+    """Signed-zero fix must not regress NaN propagation."""
+    nan = float("nan")
+    a = torch.tensor([nan, 0.0, -0.0, nan], dtype=dtype, device="cuda")
+    b = torch.tensor([-0.0, nan, nan, 0.0], dtype=dtype, device="cuda")
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_binary_arith.py]: Addressed in fa49de6. Tests now use mixed NaN/non-NaN inputs with explicit sign-bit assertions. See commit fa49de6.

File: tests/ops/test_binary_arith.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_binary_arith.py]: Addressed in fa49de6. Tests now use mixed NaN/non-NaN inputs with explicit sign-bit assertions. See commit fa49de6.

File: tileops/kernels/elementwise.py [hunk 6]
```diff
@@ -705,41 +705,55 @@ def lerp_func(a, b):
...
-        return nan_result
+        a_f32 = T.Cast("float32", a)
+        b_f32 = T.Cast("float32", b)
+        a_is_nan = T.isnan(a_f32)
+        b_is_nan = T.isnan(b_f32)
+        # Signed-zero tie-break: when a == b, prefer the operand whose sign
+        # is non-negative (+0 over -0). copysign(1, x) is +1 for +0 and -1
+        # for -0, so a_sign >= b_sign selects the positive-signed operand.
+        a_sign = T.copysign(T.FloatImm("float32", 1.0), a_f32)
+        b_sign = T.copysign(T.FloatImm("float32", 1.0), b_f32)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/elementwise.py]: Addressed in 2913cb2.

Evidence:
- Replaced `T.FloatImm("float32", 1.0)` with `one_f32 = T.cast(1.0, "float32")` and reused `one_f32` for both copysign calls
- Now consistent with all other kernels in the file (GeluKernel, LerpKernel, etc.)

Current HEAD: 2913cb2

File: tileops/kernels/elementwise.py [hunk 7]
```diff
@@ -705,41 +705,55 @@ def lerp_func(a, b):
...
-        return nan_result
+        a_f32 = T.Cast("float32", a)
+        b_f32 = T.Cast("float32", b)
+        a_is_nan = T.isnan(a_f32)
+        b_is_nan = T.isnan(b_f32)
+        # Signed-zero tie-break: when a == b, prefer the operand whose sign
+        # is negative (-0 over +0). copysign(1, x) is -1 for -0 and +1 for
+        # +0, so a_sign <= b_sign selects the negative-signed operand.
+        a_sign = T.copysign(T.FloatImm("float32", 1.0), a_f32)
+        b_sign = T.copysign(T.FloatImm("float32", 1.0), b_f32)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/elementwise.py]: Addressed in 2913cb2.

Evidence:
- Same fix applied: `one_f32 = T.cast(1.0, "float32")` with shared variable for both copysign calls
- Consistent with MaximumKernel and all other kernels in the file

Current HEAD: 2913cb2

File: tileops/kernels/elementwise.py [hunk 8]
```diff
@@ -705,41 +705,55 @@ def lerp_func(a, b):
...
+        a_is_nan = T.isnan(a_f32)
+        b_is_nan = T.isnan(b_f32)
+        # Signed-zero tie-break: when a == b, prefer the operand whose sign
+        # is non-negative (+0 over -0). copysign(1, x) is +1 for +0 and -1
+        # for -0, so a_sign >= b_sign selects the positive-signed operand.
+        a_sign = T.copysign(T.FloatImm("float32", 1.0), a_f32)
+        b_sign = T.copysign(T.FloatImm("float32", 1.0), b_f32)
+        ordered_max = T.if_then_else(
+            a > b, a, T.if_then_else(b > a, b,
+                                     T.if_then_else(a_sign >= b_sign, a, b)))
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/elementwise.py:731]: Reviewed at 2913cb2. The current implementation and the suggested alternative are functionally equivalent.

Evidence:
- Current: `(a > b) -> a, else (b > a) -> b, else tie-break` 
- Suggested: `(a == b) -> tie-break, else (a > b) -> a, else b`
- Both produce identical results. The current form is a deliberate choice: it short-circuits the copysign tie-break for the common case (unequal values) without an explicit equality check, which maps naturally to the chained comparison pattern used elsewhere in TileLang IR.
- 101/101 tests pass including 12 dedicated signed-zero tests verifying correctness.

Current HEAD: 2913cb2

File: tileops/kernels/elementwise.py [hunk 9]
```diff
@@ -705,41 +705,55 @@ def lerp_func(a, b):
...
+        a_is_nan = T.isnan(a_f32)
+        b_is_nan = T.isnan(b_f32)
+        # Signed-zero tie-break: when a == b, prefer the operand whose sign
+        # is negative (-0 over +0). copysign(1, x) is -1 for -0 and +1 for
+        # +0, so a_sign <= b_sign selects the negative-signed operand.
+        a_sign = T.copysign(T.FloatImm("float32", 1.0), a_f32)
+        b_sign = T.copysign(T.FloatImm("float32", 1.0), b_f32)
+        ordered_min = T.if_then_else(
+            a < b, a, T.if_then_else(b < a, b,
+                                     T.if_then_else(a_sign <= b_sign, a, b)))
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/elementwise.py:759]: Reviewed at 2913cb2. Same analysis as the MaximumKernel equivalent.

Evidence:
- Current and suggested forms are functionally equivalent for ordered_min
- The current chained-comparison pattern `(a < b) -> a, (b < a) -> b, else tie-break` avoids an explicit equality check and is consistent with MaximumKernel
- 101/101 tests pass including dedicated signed-zero and NaN propagation coverage.

Current HEAD: 2913cb2

**Review by xysmlx — APPROVED** [review]
